src/foam/dao/MDAO.java

```diff
@@ -82,6 +82,7 @@ public FObject put_(X x, FObject obj) {
       }
 
       state_ = index_.put(state_, obj);
+      onPut(obj);
```

Comment:
Can we move this outside the writeLock?  Don't need to block other writers while listeners are firing.

Reply:
That would require moving the return statement outside of the lock as well. I'm not too dialled into the functionality of the MDAO so I don't know if moving the return values outside of the lock will have an unintended side effect

Reply:
onPut() and return could both be moved outside the lock.

Reply:
Done